README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# gridlayout

<!-- badges: start -->

[![R-CMD-check](https://github.com/rstudio/gridlayout/workflows/R-CMD-check/badge.svg)](https://github.com/rstudio/gridlayout/actions)
<!-- badges: end -->

Build dashboard-style layouts for Shiny and RMarkdown easily using
CSS-Grid.

## Installation

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("rstudio/gridlayout")
```

## Setting up your `gridlayout`

The easiest and most common way to specify a grid layout is using an
character vector syntax where the elements in your layout are visually
lined up making the general gist of the layout clear at a glance.

``` r
library(gridlayout)

my_layout <- new_gridlayout(c(
  "      120px   1fr    1fr   ",
  "100px header  header header",
  "1fr   sidebar plot_a plot_c",
  "1fr   sidebar plot_b plot_b"
))

my_layout
#> gridlayout of 5 elements: 
#>         120px   1fr    1fr   
#>   100px header  header header
#>   1fr   sidebar plot_a plot_c
#>   1fr   sidebar plot_b plot_b
#> Gap of 12px. Total height of viewport.
#> 
#> Alternate layouts:  
#>   
#>   - Width < 500px 
#>            1fr    
#>     85px  header 
#>     350px sidebar
#>     350px plot_a 
#>     350px plot_b 
#>     350px plot_c 
#>   Gap of 12px. Total height of auto.
```

For more info and alternative ways of defining a layout see
`vignette("defining-a-layout", package = "gridlayout")`.

## Using in a shiny app

Once you’ve setup your layout, the easiest way you can use it in your
shiny apps is with the `grid_page()` ui function:

``` r
library(shiny)
library(bslib)

# The classic Geyser app with grid layout
shinyApp(
  ui = grid_page(
    layout = c(
      "     200px   1fr   ",
      "85px header  header",
      "1fr  sidebar plot  "
    ),
    grid_card_text("header", "Geysers!", is_title = TRUE),
    grid_card(
      "sidebar",
      card_header("Settings"),
      sliderInput("bins","Number of bins:", 
                  min = 1, max = 50, value = 30, width = "100%")
    ),
    grid_card(
      "plot",
      card_body_fill(
        plotOutput("distPlot")
      )
    )
  ),
  server = function(input, output) {
    output$distPlot <- renderPlot({
      x    <- faithful[, 2]
      bins <- seq(min(x), max(x), length.out = input$bins + 1)
      hist(x, breaks = bins, col = 'darkgray', border = 'white')
    })
  }
)
```

<img src="man/figures/geyser_demo.png" width="100%" /> *Screenshot of
grided geyser app running*

## Other ways of using `gridlayout` in your app

`grid_page()` will automatically make your gridlayout fill the entire
page. If you are interested in having a finer-grain control over the
size and position of your grid layout you can use the `grid_container()`
function to place your grid layout wherever you want. The equivalent app
to above can be created by replacing the UI definition with a
`fluidPage` containing a `grid_container()`:

``` r
...
shinyApp(
  ui = fluidPage(
    grid_container(
      layout = c(
        "     200px   1fr   ",
        "85px header  header",
        "1fr  sidebar plot  "
      ),
      grid_card_text("header", "Geysers!"),
      grid_card(
        "sidebar",
        card_header("Settings"),
        sliderInput("bins", "Number of bins:", 
                    min = 1, max = 50, value = 30, width = "100%")
      ),
      grid_card(
        "plot",
        card_body_fill(
          plotOutput("distPlot")
        )
      )
    )
  ),
  server = ...
)
```

This time, however the grid is constrained to `800px` tall, no-matter
how large or small the window viewing it is.

Alternatively you can use `grid_nested()` to use a `gridlayout` layout
within a panel another `gridlayout`.

``` r
...
shinyApp(
  ui = grid_page(
  layout = c(
    "     250px   1fr   ",
    "50px header  header",
    "1fr  sidebar plots "
  ),
  grid_card_text("header", "This is my header"),
  grid_card(
    "sidebar",
    card_header("Settings"),
    sliderInput("bins", "Number of bins:", min = 1, max = 50, value = 30, width = "100%")
  ),
  grid_nested(
    "plots",
    title = "Plots - in technicolor",
    layout = c(
      "distPlot distPlot distPlot",
      "redPlot  bluePlot greenPlot"
    ),
    grid_card_plot("distPlot"),
    grid_card_plot("redPlot"),
    grid_card_plot("bluePlot"),
    grid_card_plot("greenPlot")
  ),
  server = ...
)
```

## Using in RMarkdown

The function `use_gridlayout_rmd()` called in the `setup` chunk of an
RMarkdown file will enable you to use gridlayout to layout your
document. Just match the section headers to the layout element names and
place layout md table in a `gridlayout` chunk…

**`my_app.rmd`**

    ---
    title: "`gridlayout` in Rmarkdown"
    author: "Nick Strayer"
    date: "3/9/2021"
    output: html_document
    ---



    ## Main


    ```gridlayout
    |      |        |         |
    |------|--------|---------|
    |2rem  |200px   |1fr      |
    |150px |header  |header   |
    |1fr   |sidebar |main     |
    |120px |footer  |footer   |
    ```


    ## Sidebar

    Here is some content for the sidebar

    ## Footer

    Anything you want could go in the footer.

<img src="man/figures/basic_markdown.png" width="100%" />

*Output of `my_app.rmd`*


NEWS.md
# Development

### Minor new features and improvements
- Now the default container height is 100% for any container other than `grid_page()`. This helps avoid confusing sizing when layout is embedded in other containers like a card. 

# gridlayout 0.2.1

### Minor new features and improvements

- Replace internal usage of `bslib::card_body_fill()` with unsufixed `bslib::card_body()` to reflect updates in API. Will fix confusing warning messages about `card_body_fill` being depreciated even though it isn't used in user code.

### Bug fixes
- Short cards could have squished contents (15c0397483529b34688f8f76bbc57a06ac9d9764)

# gridlayout 0.2.0

### Major changes

- `grid_card()` now wraps `bslib::card()` directly. This means that there are some api changes such as depreciated the `title` argument for using `bslib::card_header(title)` instead. Old `grid_card()` arguments that are present in code will get flagged with a warning to allow users to know when/how to update their code with the new version.
- `card_plot_output()` is now depreciated in favor of the new `bslib` cards smart fill behavior. Simple plots can still be done with `grid_card_plot()`.
- The function `grid_card_old()` has been added to ease transition to the new card api and to keep some features such as collapsibility are not available on the new api.

# gridlayout 0.1.0

### Major new features and improvements

- `grid_card()` is now the main method of placing items onto the grid
- Added ability to add full-bleed header and sidebar with special child functions. These sit outside the main grid and can simplify layouts to not require them to constantly be added.
- General interface for cards is a lighter in favor of outputting most non-layout logic to the coming `bslib::card()` and friends.

### Minor new features and improvements

- Website has been updated with better resources

### Bug fixes

### Known bugs


R/grid_panel.R
#' Panel for grid page element
#'
#' Allows you to customize the appearance of an element within your gridlayout.
#' When using [grid_page] or [grid_container] all items will be automatically
#' wrapped in this function with default values. However, if further
#' customization is needed you can do this wrapping yourself. Some examples of
#' things you can customize are settings a specific element to use a given card
#' style, or to align the content of the card vertically or horizontally.
#'
#' @param area Name of grid area, should match an area defined in the layout
#'   section of the wrapping `grid_page()` or `grid_container()`.
#' @param ... Elements to include within the panel
#' @param title Character string to go across top of panel with label. If left
#'   blank the card contents will take up entire space.
#' @param use_card_styles Should the element contained in panel be made to look
#'   like an enclosed card?
#' @param collapsible Should the card be able to be collapsed (`TRUE` or
#'   `FALSE`)? Gridlayout will only show collapser if the layout allows it
#'   (panel is entirely positioned within "auto" sized rows, and has a title).
#'   Setting this to `FALSE` will mean collapsibility of the panel will never be
#'   enabled, regardless of layout.
#' @param scrollable Should scroll-bars be added so content that is larger than
#'   the panel can be seen?
#' @param v_align,h_align Vertical and Horizontal alignment of content in panel.
#'   Options include `"center", "start", "end", "stretch"`. These are a direct
#'   mapping to the the [css-spec for
#'   `justify-items`](https://developer.mozilla.org/en-US/docs/Web/CSS/justify-items)
#'    (=`h_align`) and
#'   [`align-items`](https://developer.mozilla.org/en-US/docs/Web/CSS/align-items)
#'    (= `v_align`).
#' @param panel_id ID of `grid_panel` div. Usually this is automatically set by
#'   `grid_page()` or `grid_container()`.
#' @param panel_class Class(es) of `grid_panel` div. Like `panel_id` this is
#'   typically automatically generated and should only be touched in advanced
#'   usage.
#' @param padding A valid css size that overrides the default of how much
#'   spacing is around the card contents
#'
#' @return Elements from `...` wrapped in a `shiny::div()` with `class =
#'   "grid_panel"` and vertical/horizontal alignment css applied.
#'
#' @seealso [grid_panel_text]
#' @export
#'
#' @examples
#'
#' # Simply a wrapper for shiny::div(...)
#' grid_panel(
#'   area="header",
#'   shiny::h2("R"),
#'   v_align = "center",
#'   h_align = "center"
#' )
#'
#' if (interactive()) {
#'
#' library(gridlayout)
#' library(shiny)
#'
#' # The classic Geyser app with grid layout
#' shinyApp(
#'   ui = grid_page(
#'     layout = "
#'       |2rem  |200px   |1fr    |
#'       |85px  |header  |header |
#'       |1fr   |sidebar |plot   |",
#'     grid_panel("header", h1("Geysers!")),
#'     grid_panel(
#'       "sidebar",
#'       title = "Settings",
#'       sliderInput("bins","Number of bins:", min = 1, max = 50, value = 30, width = "100%")
#'     ),
#'     grid_panel(
#'       "plot",
#'       plotOutput("distPlot", height="100%")
#'     )
#'   ),
#'   server = function(input, output) {
#'     output$distPlot <- renderPlot({
#'       x    <- faithful[, 2]
#'       bins <- seq(min(x), max(x), length.out = input$bins + 1)
#'       hist(x, breaks = bins, col = 'darkgray', border = 'white')
#'     })
#'   }
#' )
#' }
#'
grid_panel <- function(
  area,
  ...,
  v_align = NULL,
  h_align = NULL,
  title = NULL,
  use_card_styles = TRUE,
  collapsible = TRUE,
  scrollable = FALSE,
  panel_id = NULL,
  panel_class = "grid_panel card",
  padding = NULL
) {
  contents <- list(...)

  panel_styles <- htmltools::css(
    display = if (notNull(h_align) || notNull(v_align)) "grid",
    `justify-items` = if (notNull(h_align)) validate_alignment(h_align),
    `align-items` = if (notNull(v_align)) validate_alignment(v_align),
    overflow = if (scrollable) "scroll",
    `--card-padding` = if(notNull(padding)) padding
  )

  has_title <- notNull(title)
  use_collapser <- collapsible && has_title

  shiny::div(
    id = panel_id,
    class = panel_class,
    style= htmltools::css(`grid-area` = area),
    `data-gridlayout-area` = area,
    if (has_title) {
      shiny::div(
        class = "title-bar",
        shiny::h3(title),
        if (use_collapser) make_collapser_icon(),
        style = if (use_collapser) "justify-content: space-between;"
      )
    },
    shiny::div(contents, style = panel_styles, class = "panel-content")
  )
}


make_collapser_icon <- function(parent_id = "") {
  requireNamespace("fontawesome", quietly = TRUE)
  # Clicking on the collapsing icon will update classes to initiate collapsing
  # or expanding and also trigger a resize event so Shiny will know to update
  # plots that may have gotten more space after collapsing etc.
  shiny::span(
    fontawesome::fa("chevron-up"),
    "onclick" = '
    var card = this.parentElement.parentElement;
    var card_classes = card.classList;
    if (card_classes.contains("collapsed")) {
      card_classes.remove("collapsed");
    } else {
      card_classes.add("collapsed");
    }
    window.dispatchEvent(new Event("resize"));
    ',
    class = "collapser-icon"
  )
}


validate_alignment <- function(arg_val) {
  # These are just the options available to the css properties justify-self
  # (horizontal alignment) and align-self (vertical alignment)
  align_options <- c("start", "end", "center", "stretch")

  if (!arg_val %in% align_options) {
    stop("Alignment argument must be one of ", paste(align_options, collapse = ", "))
  }

  if (arg_val == "spread") {
    return ("space-evenly")
  }

  arg_val
}

el_has_class <- function(el, class_name) {
  # It's possible to have multiple "class" attributes if they were set at
  # different times. Here we just collapse all of them into one big character
  # string we can seach through

  if (!inherits(el, "shiny.tag")) return(FALSE)

  all_classes <- paste(el$attribs[names(el$attribs) == "class"], collapse = " ")

  # Look for the class name isolated from other classes. E.g if searching for
  # class of  "card"  this wont get incorectly triggered by "panel-card"
  str_detect(all_classes, paste0("(\\s|^)", class_name, "(\\s|$)"))
}

el_add_class <- function(el, class_name) {
  if (!el_has_class(el, class_name)) {
    el$attribs$class <- paste(el$attribs$class, class_name)
  }
  el
}


update_default_sized_plots <- function(el, i = "ignored") {

  using_default_settings <- identical(
    htmltools::tagGetAttribute(el, "style"),
    htmltools::tagGetAttribute(shiny::plotOutput("test"), "style")
  )
  if (using_default_settings) {
    # We give ourselves permission to modify plot styles
    el$attribs$style <- "height:100%;width:100%;min-height:150px"
  }

  el
}

